src/GafferArnold/IECoreArnoldPreview/Renderer.cpp

```diff
@@ -1933,6 +1972,197 @@ IE_CORE_FORWARDDECLARE( ArnoldObject )
 
 } // namespace
 
+//////////////////////////////////////////////////////////////////////////
+// ArnoldLightFilter
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+/// The LightFilterConnections can be used to record changes regarding connections
+/// between world space light filters and lights and to communicate them to Arnold.
+class LightFilterConnections : public IECore::RefCounted
+{
+
+public :
+
+	LightFilterConnections( bool ownConnections )
+		: m_ownConnections( ownConnections )
+	{
+	}
+
+	// Record changes. These can be called concurrently.
+	void registerLight( const std::string &lightName, ArnoldLight *light );
+	void deregisterLight( const std::string &lightName );
+	void registerLightFilter( const IECore::StringVectorData *lightNames, ArnoldLightFilter *lightFilter );
+	void deregisterLightFilter( const IECore::StringVectorData *lightNames, ArnoldLightFilter *lightFilter );
+
+	// Communicate with Arnold. This needs to get called non-concurrently before rendering.
+	void update();
+
+private:
+
+	// Mapping between many lights and many filters to avoid N*M growth in
+	// memory consumption for N lights and M filters. Every filter is put in a
+	// group that collects all filters that affect the same set of lights. Each
+	// light is then linked to all groups that affect the particular light.
+	struct FilterGroup
+	{
+		std::unordered_set<ArnoldLightFilter*> filters;
+		bool brandNew;
+		bool dirty;
+	};
+
+	using FilterGroupPtr = std::shared_ptr<FilterGroup>;
+	using FilterGroups = tbb::concurrent_hash_map<const IECore::StringVectorData*, FilterGroupPtr>;
```

Comment:
Is `StringVectorData *` the right key here? I think we're OK in terms of object identity and lifetimes because the ArnoldLightFilter class keeps the data alive via ArnoldAttributes? But for performance we're relying a lot on the Renderer client using the same exact object for each distinct set of lights aren't we? I guess Gaffer's caching of EvaluateLightLinks is definitely on our side there, but I wonder if we should use `filteredLightsData->hash()` as the key instead?

Reply:
Those were my thoughts, as well, and I concluded that things would work out alright, but I've also felt that using the raw pointer here is dodgy and will look dangerous to everyone else reading the code. I'm happy to use the hash instead. Probably worth running another performance test with the hashing in place, though?

Reply:
The hash should be cheap enough I think - it's stored internal to the StringVectorData so won't be computed repeatedly unless we have different objects. Always good to measure again though...

Reply:
I had missed that I'm actually using the pointer in `LightFilterConnections::update()` to loop through all lights that are affected by the FilterGroup. I'll keep things the way they are for now as working around this problem would probably obfuscate the code more.